snakemake/resources.py

```diff
@@ -1,5 +1,18 @@
 import re
 
+supported_keys = {
```

Comment:
What about "quality of service" (`--qos`)?

Reply:
Hi,

Please note, this is a PR Johannes opened because I was away and he wanted to comment in a systematic way. Now, that I can proceed testing (and develop further, which all will take some time), quite a few things need to be cleaned.

As to your question: No, the idea is not to mimic all sbatch options. Rather the most common ones to also catch errors and to have a catch-all option for the rest. In that sense, 'constraint' is also a doubtful candidate as I have to admit. Both, 'constraint' and 'qos' are handled differently from plugin to plugin, from cluster to cluster. This is a bit unfortunate, as is may have substantial impact on throughput. In a version 2 of this feature, we need to think about pre-rule injections, such that this can be handled without altering each workflow.

Just my .02€

Reply:
To make sure I understand: using the built-in Slurm executor implemented in this PR, I wouldn't be able to specify a different value for `--qos` for different rules. Since `qos` is not one of the supported keys, the rules below would trigger an error:

```
rule rule1:
    ...
    resources:
        qos="qos1",
    ...

rule rule2:
    ...
    resources:
        qos="qos2",
    ...

```

Reply:
Mhm, I don't see a reason why one would not want to allow that, in particular since it is now possible to define rule resources at the command line (so that the workflow does not get polluted with platform specific stuff). @cmeesters can you explain?

Reply:
I assume you are doing something fancy with the resources listed in `supported_keys`, i.e. trying to manage the requested resources with the total amount available. I think that's fine to limit this feature to ones that make sense to manage. But I don't see why you can't pass arbitrary resources directory to `sbatch`. If I write a rule definition like:

```
rule fruit:
    resources:
        mem_mb=4000,
        banana="apple",
```

Then this would result in `sbatch --mem=4000 --banana=apple`. Obviously this will cause sbatch to error, but that is on the end user. I don't think Snakemake needs to monitor all ~80 slurm options. You don't want to have to update Snakemake every time Slurm adds a new option.

Reply:
Sounds reasonable to me. @cmeesters thoughts?

Reply:
My ideal resource support would:

- allow to pass every SLURM argument, like proposed by @jdblischak .
- have standard ones, which are validated and checked for common typos.
- have standard ones, which have snakemake resemblence and allow for a snakemake->SLURM conversion. E.g. in an smp-case from cpus to cpus-per-task. 

In a first step, we could just drop the `supported_keys` alltogether and follow @jdblischak suggestion to an extreme and later add validators/translators. So, yes that is a jolly good point.

Reply:
In my fork of the slurm profile I have documented a similar approach to dealing with resources and adding a convenience parser for time as well https://github.com/mbhall88/slurm#rule-specific-resource-configuration. Not sure if this is helpful or not, but thought I'd drop it here in case it helps

Reply:
oh, I thought I had _already_ written an answer to you @mbhall88 

Well, anyway, I think this is a jolly good idea. In a discussion with @johanneskoester we agreed to add it to the executor class, respectively the resources. Johannes will approach you about this (please keep me in the loop, if you will). One question, though: Your code is covered by MIT, so taking it poses no problem. May I put your name and link to the code and refer to it appropriately?

Reply:
Absolutely. Take whatever you like.